### PR TITLE
cosmo : move func tests to own file and test verbose flag

### PR DESCRIPTION
**Pull request by nstarman**
This PR does 2 things:
1) get us to 100% codecov of cosmo/funcs.py by adding a test for the verbose flag in z_at_value.
2) Separates the test functions for func.py into their own file. Test_cosmology.py mostly tests features in core.py, so func stuff is a little out of place. And while there is currently only 1 function is func.py, it would be nice to see the cosmo package grow, so having tests mimic the structure of the code keeps stuff organized.

Labels: no-changelog-needed
Request Review: @dhomeier  

Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

### CODE REVIEW

**Comment by nstarman** [comment]
> Could you add a brief description on the motivation of this PR

Done!
> also is there an issue linked?

No specific issue. This came from a comment made in the last PR where I was spitballing how to test the verbose flag. It ended up requiring some slight modifications as Windows has odd issues with tempfile.

**Comment by mhvk** [comment]
@nstarman - I think this is a good idea, but to make review easier, could you point out where exactly a change was made (i.e., where it is not just moving the tests around)?

**Comment by mhvk** [comment]
Since this only affects tests, no need for a change log entry.

**Comment by nstarman** [comment]
> @nstarman - I think this is a good idea, but to make review easier, could you point out where exactly a change was made (i.e., where it is not just moving the tests around)?

Oh yes. The diff file isn't so useful. The added test is line 48 of test_func.py

**Comment by nstarman** [comment]
I wasn't super pleased with the need for a TemporaryDirectory with a file inside, so I looked at the documentation for file-like structures and figured out this test can be done a LOT more simply using a StringIO. I can't test the Windows without pushing, so I'll revert if it doesn't work. 🤞

**Comment by nstarman** [comment]
Hooray! The simplification works. No more temporary directories.
Change is still on line 48+.

File: astropy/cosmology/tests/test_funcs.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    # raise a CosmologyError)
+    with pytest.raises(core.CosmologyError):
+        with pytest.warns(AstropyUserWarning, match=r'fval is not bracketed'):
+            z_at_value(cosmo.angular_diameter_distance, 1500*u.Mpc, zmax=0.5)
+
+    with pytest.raises(core.CosmologyError):
+        with pytest.warns(AstropyUserWarning, match=r'fval is not bracketed'):
+            z_at_value(cosmo.angular_diameter_distance, 1500*u.Mpc, zmin=4.)
+
+    # Test the "verbose" flag. Since this uses "print", ned to mod stdout
```
Comment (mhvk) [inline astropy/cosmology/tests/test_funcs.py:48]: typo in comment

File: astropy/cosmology/tests/test_funcs.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    original_stdout = sys.stdout
```
Comment (mhvk) [inline astropy/cosmology/tests/test_funcs.py:49]: The standard pytest way to do this is using 
https://docs.pytest.org/en/stable/monkeypatch.html

If so, easiest to make this a new test function (generally, in astropy we have been putting a bit too many tests in our "unit tests"... I'm definitely guilty!)

**Review by mhvk — COMMENTED** [review]
Looks good to me, except two comments, one tiny, the other small.

File: astropy/cosmology/tests/test_funcs.py (hunk 1 above)
Comment (nstarman) [reply astropy/cosmology/tests/test_funcs.py:48]: ✅

File: astropy/cosmology/tests/test_funcs.py (hunk 2 above)
Comment (nstarman) [reply astropy/cosmology/tests/test_funcs.py:49]: Very cool pytest fixture.

**Review by mhvk — APPROVED** [review]
Great, looks all OK now.

**Pull request merged**